packages/adapters/watcher/src/verifiers/proposal.ts

```diff
@@ -0,0 +1,249 @@
+import { createLoggingContext, jsonifyError, RequestContext, domainToChainId } from "@connext/nxtp-utils";
+import { Contract, providers, utils } from "ethers";
+
+import { ProposedData, Verifier, VerifierContext, VerifyResponse } from "../types";
+import { virtualTreeBytecode } from "../bytecodes";
+import { EMPTY_ROOT } from "../helpers/constants";
+
+export class ProposedRootVerifier extends Verifier {
+  constructor(
+    context: VerifierContext,
+    private readonly hubDomain: number,
+    private readonly hubProvider: providers.JsonRpcProvider,
+  ) {
+    super(context);
+  }
+
+  /**
+   * @notice Validate whether the proposed aggregate root is correct.
+   * @dev Should compare the proposed aggregate root to the virtual aggregate root. If
+   * the invariant is violated, should return a value to indicate that the caller/consumer should escalate.
+   *
+   * @returns object - Whether the invariant was verified. `true` if propose was verified (no switch needed).
+   * `false` if the invariant was violated (switch is needed)! If false, a reason is included in the response object.
+   */
+  public override async checkInvariant(requestContext: RequestContext): Promise<VerifyResponse> {
+    // get root manager contract of the hub domain chain
+    const hubDomainChainId = domainToChainId(this.hubDomain);
+    const rootManagerContract = this.getRootManagerContract(hubDomainChainId);
+
+    // get proposed data from root manager contract
+    const proposedData = await this.getProposedData(rootManagerContract, hubDomainChainId, requestContext);
+    const { aggregateRoot: proposedRoot } = proposedData;
+
+    // stop check execution if the proposed root dispute period has elapsed
+    const isDisputeTimeOver = await this.checkDisputeTimeFinalization(+proposedData.endOfDispute);
+    if (isDisputeTimeOver) {
+      return {
+        needsAction: false,
+        reason: `Dispute time for root ${proposedRoot} is already over`,
+      };
+    }
```

Comment:
if the dispute time elapses, and there is a valid reason to error on the proposal, we should likely make sure that is handled with at least some explicit logging if not alerts